### PR TITLE
Standardize callables

### PR DESCRIPTION
**Pull request by collindutter**
- [x] I have read and agree to the contributing guidelines for [submitting new pull requests](https://github.com/griptape-ai/griptape?tab=readme-ov-file#submitting-pull-requests).

## Describe your changes
[Relevant thread](https://github.com/griptape-ai/griptape/issues/285#issuecomment-2427912216)

### Changed
- **BREAKING**: Renamed callables throughout the framework for consistency:
  - Renamed `LocalStructureRunDriver.structure_factory_fn` to `LocalStructureRunDriver.create_structure`.
  - Renamed `SnowflakeSqlDriver.connection_func` to `SnowflakeSqlDriver.get_connection`.
  - Renamed `CsvLoader.formatter_fn` to `CsvLoader.format_row`.
  - Renamed `SqlLoader.formatter_fn` to `SqlLoader.format_row`.
  - Renamed `CsvExtractionEngine.system_template_generator` to `CsvExtractionEngine.generate_system_template`.
  - Renamed `CsvExtractionEngine.user_template_generator` to `CsvExtractionEngine.generate_user_template`.
  - Renamed `JsonExtractionEngine.system_template_generator` to `JsonExtractionEngine.generate_system_template`.
  - Renamed `JsonExtractionEngine.user_template_generator` to `JsonExtractionEngine.generate_user_template`.
  - Renamed `PromptResponseRagModule.generate_system_template` to `PromptResponseRagModule.generate_system_template`.
  - Renamed `PromptTask.generate_system_template` to `PromptTask.generate_system_template`.
  - Renamed `ToolkitTask.generate_assistant_subtask_template` to `ToolkitTask.generate_assistant_subtask_template`.
  - Renamed `JsonSchemaRule.template_generator` to `JsonSchemaRule.generate_template`.
  - Renamed `ToolkitTask.generate_user_subtask_template` to `ToolkitTask.generate_user_subtask_template`.
  - Renamed `TextLoaderRetrievalRagModule.process_query_output_fn` to `TextLoaderRetrievalRagModule.process_query_output`.
  - Renamed `FuturesExecutorMixin.futures_executor_fn` to `FuturesExecutorMixin.create_futures_executor`.
  - Renamed `VectorStoreTool.process_query_output_fn` to `VectorStoreTool.process_query_output`.
  - Renamed `CodeExecutionTask.run_fn` to `CodeExecutionTask.on_run`.
  - Renamed `Chat.input_fn` to `Chat.handle_input`.
  - Renamed `Chat.output_fn` to `Chat.handle_output`.
  - Renamed `EventListener.handler` to `EventListener.on_event`.

## Issue ticket number and link
Closes https://github.com/griptape-ai/griptape/issues/285

----
📚 Documentation preview 📚: https://griptape--1275.org.readthedocs.build//1275/

### CODE REVIEW

**Comment by vachillo** [comment]
what is the reasoning between `factory` and `provider`?

**Comment by collindutter** [comment]
Reposted my comment linked in the description.

>*_factory: Used for functions that create a new instance of an object.
*_provider: Used for functions that supply a value. These functions may create new instances, as long as the value represents the same resource or value.

Multiple calls to a factory should give new objects with new meaning each time. For instance, 3 calls to `structure_factory` should give me 3 structures that I should treat as independent. 3 calls to `user_template_provider` may give me 3 `J2` objects, but they represent the same thing under the hood.

File: griptape/drivers/sql/snowflake_sql_driver.py [hunk 1]
```diff
@@ -15,24 +15,26 @@
 
 @define
 class SnowflakeSqlDriver(BaseSqlDriver):
-    connection_func: Callable[[], SnowflakeConnection] = field(kw_only=True)
+    connection_provider: Callable[[], SnowflakeConnection] = field(kw_only=True)
```
Comment (dylanholmes) [inline griptape/drivers/sql/snowflake_sql_driver.py:18]: Might be nice to create some reusable types for these to make it easier for customer code create their own abstractions.

For example adding a:

```py
ConnectionProvider = Callable[[], SnowflakeConnection]
```

In this file, then using it here:


```py
connection_provider: ConnectionProvider = field(kw_only=True)
```

And you can reduce duplicate in `def validate_connection_provider(` too.

I can also see arguments against this, but I like being able to easily references types in my own code.

File: griptape/drivers/vector/local_vector_store_driver.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
 class LocalVectorStoreDriver(BaseVectorStoreDriver):
     entries: dict[str, BaseVectorStoreDriver.Entry] = field(factory=dict)
     persist_file: Optional[str] = field(default=None)
-    relatedness_fn: Callable = field(default=lambda x, y: dot(x, y) / (norm(x) * norm(y)))
+    relatedness_calculator: Callable = field(default=lambda x, y: dot(x, y) / (norm(x) * norm(y)))
```
Comment (dylanholmes) [inline griptape/drivers/vector/local_vector_store_driver.py:22]: Why no generic params?

File: tests/integration/tasks/test_text_summary_task.py [hunk 3]
```diff
@@ -7,7 +7,7 @@ class TestTextSummaryTask:
     @pytest.fixture(
         autouse=True,
         params=StructureTester.TEXT_SUMMARY_TASK_CAPABLE_PROMPT_DRIVERS,
-        ids=StructureTester.prompt_driver_id_fn,
+        ids=StructureTester.generate_prompt_driver_id,
```
Comment (dylanholmes) [inline tests/integration/tasks/test_text_summary_task.py:10]: Why isn't this `prompt_driver_id_provider` or `prompt_driver_id_factory`?

**Review by dylanholmes — APPROVED** [review]
Nice work! Like the renames.

Just optional suggestions. Don't necessarily need to be done in this PR either.

File: griptape/drivers/sql/snowflake_sql_driver.py (hunk 1 above)
Comment (collindutter) [reply griptape/drivers/sql/snowflake_sql_driver.py:18]: yeah I like the idea, but that can be introduced in the future.

File: griptape/drivers/vector/local_vector_store_driver.py (hunk 2 above)
Comment (collindutter) [reply griptape/drivers/vector/local_vector_store_driver.py:22]: They should be there but I want to keep this PR strictly on naming.

File: tests/integration/tasks/test_text_summary_task.py (hunk 3 above)
Comment (collindutter) [reply tests/integration/tasks/test_text_summary_task.py:10]: Because that's actually a class method, not a field. Methods should still be verb-noun.

**Review by vachillo — DISMISSED** [review]

**Review by vachillo — APPROVED** [review]
lets update the PR description for posterity

**Pull request merged**